src/azul/chalice.py

```diff
@@ -146,21 +146,25 @@ def spec(self) -> JSON:
         }
 
     def self_url(self, endpoint_path=None) -> str:
-        self_url = furl(config.service_endpoint())
-        # self.current_request is instance of Request during AWS Lambda and
-        # ChaliceLocal executions
-        if isinstance(self.current_request, Request):
+        self_url = furl()
```

Comment:
It's stronger to assign in each branch. That enables a PyCharm check for uninitialized variable.

Apply this patch:

```
Index: src/azul/chalice.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/azul/chalice.py b/src/azul/chalice.py
--- a/src/azul/chalice.py	(revision 73859bb58688e5788c8f6306f43f3ab3cbd8ebfb)
+++ b/src/azul/chalice.py	(date 1623109987363)
@@ -146,25 +146,24 @@
         }
 
     def self_url(self, endpoint_path=None) -> str:
-        self_url = furl()
-        # self.current_request is None during AWS StepFunction executions
         if self.current_request is None:
-            self_url.url = config.service_endpoint()
-        # and instance of Request during AWS Lambda and ChaliceLocal executions
+            # Invocation via AWS StepFunctions
+            self_url = furl(config.service_endpoint())
         elif isinstance(self.current_request, Request):
             try:
                 scheme = self.current_request.headers['x-forwarded-proto']
             except KeyError:
-                # Only allow `http` in `make local`
+                # Invocation via `chalice local` or tests 
                 assert self.lambda_context.function_name == 'api_handler'
-                scheme = 'http'
+                scheme = 'http' 
             else:
+                # Invocation via API Gateway 
                 assert self.lambda_context.function_name == config.service_name
-            self_url.set(scheme=scheme, netloc=self.current_request.headers['host'])
-            if endpoint_path is None:
-                endpoint_path = self.current_request.context['path']
+            self_url = furl(scheme=scheme, netloc=self.current_request.headers['host'])
         else:
             assert False, self.current_request
+        if endpoint_path is None:
+            endpoint_path = self.current_request.context['path']
         return self_url.set(path=endpoint_path).url
 
     def _register_spec(self,
```

remove the else branch, and notice the PyCharm warning:

![image](https://user-images.githubusercontent.com/5143256/121101685-2c38f080-c7b1-11eb-97bc-c6e6403c92fc.png)